app/services/create_subscription_service.rb

```diff
@@ -0,0 +1,39 @@
+class CreateSubscriptionService < ApplicationService
+  attr_reader :subscriber_list, :subscriber, :frequency, :current_user
+
+  def initialize(subscriber_list, subscriber, frequency, current_user)
+    @subscriber_list = subscriber_list
+    @subscriber = subscriber
+    @frequency = frequency
+    @current_user = current_user
+  end
+
+  def call
+    ApplicationRecord.transaction do
+      subscriber.lock!
+
+      subscription = Subscription.find_by(
+        subscriber_list: subscriber_list,
+        subscriber: subscriber,
+      )
+
+      if subscription
+        return subscription if subscription.frequency == frequency
```

Comment:
it would be fine to maintain the 200/201 api interface by returning something like `[:found, subscription]`/`[:created, subscription]` from this service (or some other approach). Would we want to use the knowledge of create vs update at some point?

Reply:
Yeah, I did wonder about returning some kind of flag - we could also return the previous subscription (if any), so the service doesn't need to know how it's used.

I decided to go the other way and question: why does it matter, as long as we do what was requested? If we do want to use this knowledge in future, then we could always add the differing response codes back in. However, we generally mask these kinds of nuances from the user - in this case, we always say "[you've subscribed](https://github.com/alphagov/email-alert-frontend/blob/master/app/controllers/subscription_authentication_controller.rb#L28)", irrespective of how many times they click the link to confirm their subscription.

Reply:
Cool, I guess if we've taken that as a firm decision, then this seems fine 👍